feat(next): Add ability to dynamically compute iron options based on req, res

searchableguy opened the pull request:
This adds ability to pass options built using request or response object. 

This will resolve the issues people have in https://github.com/vvo/iron-session/issues/410

@vvo mentioned making maxCookieAge or cookieName functions but that approach will require changing the signature of the underlying cookie type which is coming from another package. It is also not very flexible and will require bigger changes. 

That's why I'm going with current approach where you can pass a function which returns option object and use request or response props in it. 

I'm happy to make any changes. Let me know.

vvo commented:
Hey there this is a good approach, can you:
1. Explain why you would need this method async and not sync?
2. Add simple tests
3. Add documentation
?

Thanks!

searchableguy commented:
@vvo 
The main reason for having the async signature is if someone wants to call another async function inside and wait for its results. For my purpose, I need to make a network call to get the maxCookieAge hence it is async for ease of use. 

I can write type signature to allow for both sync and async function. 

Awaiting a sync function wouldn't change the result so it will work the same.

I will write some tests and docs. Do I change the package version as well or leave it to you?

searchableguy commented:
Hello @vvo I added tests, wrote about the presence of this functionality and example in the README.md, fixed type signature to allow for both sync and async. 

Anything else you need from my end?

vvo commented:
> I will write some tests and docs. Do I change the package version as well or leave it to you?

I will do the release, thanks.

> Anything else you need from my end?

Can you add the req and res types in the API docs too?

And we will be good to go.

searchableguy commented:
Done.

searchableguy commented:
My changes won't affect them. So if it's purely about API consistency, then yes.

vvo commented:
> So if it's purely about API consistency, then yes but nobody has opened an issue about it AFAIK so I'm not sure if people want it.

I am fine waiting for someone to request that and see how it goes then.

searchableguy commented:
I think you may need to approve the vercel run request for kickstarting the tests and checks.

vvo commented:
@all-contributors please add @searchableguy tests, doc and code
